File: internal/models/stock.go
package models

import (
	"encoding/json"
	"time"

	"github.com/shopspring/decimal"
)

// JSONDecimal 用于JSON序列化的decimal类型，确保序列化为数字而不是字符串
type JSONDecimal struct {
	decimal.Decimal
}

// MarshalJSON 实现JSON序列化，将decimal转换为float64
func (d JSONDecimal) MarshalJSON() ([]byte, error) {
	f, _ := d.Float64()
	return json.Marshal(f)
}

// UnmarshalJSON 实现JSON反序列化
func (d *JSONDecimal) UnmarshalJSON(data []byte) error {
	var f float64
	if err := json.Unmarshal(data, &f); err != nil {
		// 如果解析float64失败，尝试解析字符串
		var s string
		if err2 := json.Unmarshal(data, &s); err2 != nil {
			return err
		}
		dec, err3 := decimal.NewFromString(s)
		if err3 != nil {
			return err3
		}
		d.Decimal = dec
		return nil
	}
	d.Decimal = decimal.NewFromFloat(f)
	return nil
}

// NewJSONDecimal 创建新的JSONDecimal
func NewJSONDecimal(d decimal.Decimal) JSONDecimal {
	return JSONDecimal{d}
}

// StockBasic 股票基本信息
type StockBasic struct {
	TSCode   string `json:"ts_code"`   // 股票代码
	Symbol   string `json:"symbol"`    // 股票简称
	Name     string `json:"name"`      // 股票名称
	Area     string `json:"area"`      // 所在地域
	Industry string `json:"industry"`  // 所属行业
	Market   string `json:"market"`    // 市场类型
	ListDate string `json:"list_date"` // 上市日期
}

// StockDaily 股票日线数据
type StockDaily struct {
	TSCode    string      `json:"ts_code"`    // 股票代码
	TradeDate string      `json:"trade_date"` // 交易日期 YYYYMMDD
	Open      JSONDecimal `json:"open"`       // 开盘价
	High      JSONDecimal `json:"high"`       // 最高价
	Low       JSONDecimal `json:"low"`        // 最低价
	Close     JSONDecimal `json:"close"`      // 收盘价
	PreClose  JSONDecimal `json:"pre_close"`  // 昨收价
	Change    JSONDecimal `json:"change"`     // 涨跌额
	PctChg    JSONDecimal `json:"pct_chg"`    // 涨跌幅
	Vol       JSONDecimal `json:"vol"`        // 成交量(手)
	Amount    JSONDecimal `json:"amount"`     // 成交额(千元)
}

// TechnicalIndicators 技术指标
type TechnicalIndicators struct {
	TSCode    string                   `json:"ts_code"`
	TradeDate string                   `json:"trade_date"`
	MACD      *MACDIndicator           `json:"macd,omitempty"`
	RSI       *RSIIndicator            `json:"rsi,omitempty"`
	BOLL      *BollingerBandsIndicator `json:"boll,omitempty"`
	MA        *MovingAverageIndicator  `json:"ma,omitempty"`
	KDJ       *KDJIndicator            `json:"kdj,omitempty"`
	// 新增动量因子
	WR       *WilliamsRIndicator `json:"wr,omitempty"`
	Momentum *MomentumIndicator  `json:"momentum,omitempty"`
	ROC      *ROCIndicator       `json:"roc,omitempty"`
	// 新增趋势因子
	ADX      *ADXIndicator      `json:"adx,omitempty"`
	SAR      *SARIndicator      `json:"sar,omitempty"`
	Ichimoku *IchimokuIndicator `json:"ichimoku,omitempty"`
	// 新增波动率因子
	ATR    *ATRIndicator                  `json:"atr,omitempty"`
	StdDev *StdDevIndicator               `json:"stddev,omitempty"`
	HV     *HistoricalVolatilityIndicator `json:"hv,omitempty"`
	// 新增成交量因子
	VWAP   *VWAPIndicator   `json:"vwap,omitempty"`
	ADLine *ADLineIndicator `json:"ad_line,omitempty"`
	EMV    *EMVIndicator    `json:"emv,omitempty"`
	VPT    *VPTIndicator    `json:"vpt,omitempty"`
}

// MACDIndicator MACD指标
type MACDIndicator struct {
	DIF       JSONDecimal `json:"dif"`       // DIF线
	DEA       JSONDecimal `json:"dea"`       // DEA线(信号线)
	Histogram JSONDecimal `json:"histogram"` // MACD柱状图
	Signal    string      `json:"signal"`    // 金叉/死叉信号
}

// RSIIndicator RSI指标
type RSIIndicator struct {
	RSI14  JSONDecimal `json:"rsi14"`  // 14日RSI
	Signal string      `json:"signal"` // 超买/超卖信号
}

// BollingerBandsIndicator 布林带指标
type BollingerBandsIndicator struct {
	Upper  JSONDecimal `json:"upper"`  // 上轨
	Middle JSONDecimal `json:"middle"` // 中轨(MA20)
	Lower  JSONDecimal `json:"lower"`  // 下轨
	Signal string      `json:"signal"` // 突破信号
}

// MovingAverageIndicator 移动平均线指标
type MovingAverageIndicator struct {
	MA5   JSONDecimal `json:"ma5"`   // 5日均线
	MA10  JSONDecimal `json:"ma10"`  // 10日均线
	MA20  JSONDecimal `json:"ma20"`  // 20日均线
	MA60  JSONDecimal `json:"ma60"`  // 60日均线
	MA120 JSONDecimal `json:"ma120"` // 120日均线
}

// KDJIndicator KDJ指标
type KDJIndicator struct {
	K      JSONDecimal `json:"k"`      // K值
	D      JSONDecimal `json:"d"`      // D值
	J      JSONDecimal `json:"j"`      // J值
	Signal string      `json:"signal"` // 买卖信号
}

// PredictionResult 预测结果
type PredictionResult struct {
	TSCode      string                   `json:"ts_code"`
	TradeDate   string                   `json:"trade_date"`
	Predictions []TradingPointPrediction `json:"predictions"`
	Confidence  JSONDecimal              `json:"confidence"` // 预测置信度
	UpdatedAt   time.Time                `json:"updated_at"`
}

// TradingPointPrediction 买卖点预测
type TradingPointPrediction struct {
	Type           string      `json:"type"`             // "BUY" 或 "SELL"
	Price          JSONDecimal `json:"price"`            // 预测价格
	Date           string      `json:"date"`             // 预测日期
	Probability    JSONDecimal `json:"probability"`      // 概率
	Reason         string      `json:"reason"`           // 预测理由
	Indicators     []string    `json:"indicators"`       // 相关指标
	SignalDate     string      `json:"signal_date"`      // 信号产生的日期（基于哪一天的数据）
	Backtested     bool        `json:"backtested"`       // 是否已回测
	IsCorrect      bool        `json:"is_correct"`       // 预测是否正确
	NextDayPrice   JSONDecimal `json:"next_day_price"`   // 第二天收盘价
	PriceDiff      JSONDecimal `json:"price_diff"`       // 价格差值
	PriceDiffRatio JSONDecimal `json:"price_diff_ratio"` // 价格差值百分比
}

// APIResponse 通用API响应结构
type APIResponse struct {
	Success bool        `json:"success"`
	Message string      `json:"message,omitempty"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
}

// FavoriteStock 收藏股票结构
type FavoriteStock struct {
	ID        string    `json:"id"`         // 唯一标识
	TSCode    string    `json:"ts_code"`    // 股票代码
	Name      string    `json:"name"`       // 股票名称
	StartDate string    `json:"start_date"` // 收藏时的开始日期
	EndDate   string    `json:"end_date"`   // 收藏时的结束日期
	GroupID   string    `json:"group_id"`   // 所属分组ID
	SortOrder int       `json:"sort_order"` // 排序顺序
	CreatedAt time.Time `json:"created_at"` // 创建时间
	UpdatedAt time.Time `json:"updated_at"` // 更新时间
}

// FavoriteGroup 收藏分组结构
type FavoriteGroup struct {
	ID        string    `json:"id"`         // 唯一标识
	Name      string    `json:"name"`       // 分组名称
	Color     string    `json:"color"`      // 分组颜色
	SortOrder int       `json:"sort_order"` // 排序顺序
	CreatedAt time.Time `json:"created_at"` // 创建时间
	UpdatedAt time.Time `json:"updated_at"` // 更新时间
}

// FavoriteStockRequest 添加收藏股票请求
type FavoriteStockRequest struct {
	TSCode    string `json:"ts_code"`    // 股票代码
	Name      string `json:"name"`       // 股票名称
	StartDate string `json:"start_date"` // 开始日期
	EndDate   string `json:"end_date"`   // 结束日期
	GroupID   string `json:"group_id"`   // 所属分组ID
}

// UpdateFavoriteRequest 更新收藏股票请求
type UpdateFavoriteRequest struct {
	StartDate string `json:"start_date"` // 开始日期
	EndDate   string `json:"end_date"`   // 结束日期
	GroupID   string `json:"group_id"`   // 所属分组ID
	SortOrder int    `json:"sort_order"` // 排序顺序
}

// CreateGroupRequest 创建分组请求
type CreateGroupRequest struct {
	Name  string `json:"name"`  // 分组名称
	Color string `json:"color"` // 分组颜色
}

// UpdateGroupRequest 更新分组请求
type UpdateGroupRequest struct {
	Name      string `json:"name"`       // 分组名称
	Color     string `json:"color"`      // 分组颜色
	SortOrder int    `json:"sort_order"` // 排序顺序
}

// UpdateFavoritesOrderRequest 更新收藏排序请求
type UpdateFavoritesOrderRequest struct {
	FavoriteOrders []FavoriteOrderItem `json:"favorite_orders"` // 收藏排序列表
}

// FavoriteOrderItem 收藏排序项
type FavoriteOrderItem struct {
	ID        string `json:"id"`         // 收藏ID
	GroupID   string `json:"group_id"`   // 所属分组ID
	SortOrder int    `json:"sort_order"` // 排序顺序
}

// HealthStatus 健康检查状态
type HealthStatus struct {
	Status    string            `json:"status"`
	Timestamp time.Time         `json:"timestamp"`
	Version   string            `json:"version"`
	Services  map[string]string `json:"services"`
}

// SimpleIndicator 简单技术指标
type SimpleIndicator struct {
	MA5            string `json:"ma5"`
	MA10           string `json:"ma10"`
	MA20           string `json:"ma20"`
	CurrentPrice   string `json:"current_price"`
	PriceChange    string `json:"price_change"`
	PriceChangePct string `json:"price_change_pct"`
	Trend          string `json:"trend"`
	LastUpdate     string `json:"last_update"`
}

// FavoriteSignal 收藏股票信号
type FavoriteSignal struct {
	ID           string          `json:"id"`
	TSCode       string          `json:"ts_code"`
	Name         string          `json:"name"`
	GroupID      string          `json:"group_id"`
	CurrentPrice string          `json:"current_price"`
	TradeDate    string          `json:"trade_date"`
	Indicators   SimpleIndicator `json:"indicators"`
	Predictions  interface{}     `json:"predictions"` // 保持interface{}因为预测结构可能变化
	UpdatedAt    string          `json:"updated_at"`
}

// FavoritesSignalsResponse 收藏股票信号响应
type FavoritesSignalsResponse struct {
	Total             int              `json:"total"`
	Signals           []FavoriteSignal `json:"signals"`
	Calculating       bool             `json:"calculating"`
	CalculationStatus interface{}      `json:"calculation_status,omitempty"`
}

// CandlestickPattern 蜡烛图模式识别结果
type CandlestickPattern struct {
	TSCode      string      `json:"ts_code"`      // 股票代码
	TradeDate   string      `json:"trade_date"`   // 交易日期
	Pattern     string      `json:"pattern"`      // 图形名称
	Signal      string      `json:"signal"`       // 买卖信号
	Confidence  JSONDecimal `json:"confidence"`   // 置信度
	Description string      `json:"description"`  // 图形描述
	Strength    string      `json:"strength"`     // 信号强度 (STRONG, MEDIUM, WEAK)
	Volume      JSONDecimal `json:"volume"`       // 成交量
	PriceChange JSONDecimal `json:"price_change"` // 价格变化
}

// VolumePricePattern 量价图形识别结果
type VolumePricePattern struct {
	TSCode      string      `json:"ts_code"`      // 股票代码
	TradeDate   string      `json:"trade_date"`   // 交易日期
	Pattern     string      `json:"pattern"`      // 图形名称
	Signal      string      `json:"signal"`       // 买卖信号
	Confidence  JSONDecimal `json:"confidence"`   // 置信度
	Description string      `json:"description"`  // 图形描述
	Strength    string      `json:"strength"`     // 信号强度
	Volume      JSONDecimal `json:"volume"`       // 成交量
	PriceChange JSONDecimal `json:"price_change"` // 价格变化
	VolumeRatio JSONDecimal `json:"volume_ratio"` // 量比
}

// PatternRecognitionResult 图形识别结果
type PatternRecognitionResult struct {
	TSCode            string               `json:"ts_code"`            // 股票代码
	TradeDate         string               `json:"trade_date"`         // 交易日期
	Candlestick       []CandlestickPattern `json:"candlestick"`        // 蜡烛图模式
	VolumePrice       []VolumePricePattern `json:"volume_price"`       // 量价图形
	CombinedSignal    string               `json:"combined_signal"`    // 综合信号
	OverallConfidence JSONDecimal          `json:"overall_confidence"` // 综合置信度
	RiskLevel         string               `json:"risk_level"`         // 风险等级
}

// PatternSearchRequest 图形搜索请求
type PatternSearchRequest struct {
	TSCode        string   `json:"ts_code"`        // 股票代码
	StartDate     string   `json:"start_date"`     // 开始日期
	EndDate       string   `json:"end_date"`       // 结束日期
	Patterns      []string `json:"patterns"`       // 要搜索的图形类型
	MinConfidence float64  `json:"min_confidence"` // 最小置信度
}

// PatternSearchResponse 图形搜索响应
type PatternSearchResponse struct {
	Total   int                        `json:"total"`
	Results []PatternRecognitionResult `json:"results"`
}

// PatternSummary 图形模式摘要
type PatternSummary struct {
	TSCode    string         `json:"ts_code"`    // 股票代码
	Period    int            `json:"period"`     // 统计周期（天数）
	StartDate string         `json:"start_date"` // 开始日期
	EndDate   string         `json:"end_date"`   // 结束日期
	Patterns  map[string]int `json:"patterns"`   // 各种图形模式统计
	Signals   map[string]int `json:"signals"`    // 各种信号统计
	UpdatedAt time.Time      `json:"updated_at"` // 更新时间
}

// RecentSignal 最近的图形信号
type RecentSignal struct {
	TSCode      string      `json:"ts_code"`     // 股票代码
	TradeDate   string      `json:"trade_date"`  // 交易日期
	Pattern     string      `json:"pattern"`     // 图形名称
	Signal      string      `json:"signal"`      // 买卖信号
	Confidence  JSONDecimal `json:"confidence"`  // 置信度
	Description string      `json:"description"` // 图形描述
	Strength    string      `json:"strength"`    // 信号强度
	Type        string      `json:"type"`        // 信号类型
}

// StockSignal 股票信号存储结构
type StockSignal struct {
	ID                  string                     `json:"id"`                             // 唯一标识
	TSCode              string                     `json:"ts_code"`                        // 股票代码
	Name                string                     `json:"name"`                           // 股票名称
	TradeDate           string                     `json:"trade_date"`                     // 信号基于的交易日期
	SignalDate          string                     `json:"signal_date"`                    // 信号计算日期
	SignalType          string                     `json:"signal_type"`                    // 信号类型: BUY, SELL, HOLD
	SignalStrength      string                     `json:"signal_strength"`                // 信号强度: STRONG, MEDIUM, WEAK
	Confidence          JSONDecimal                `json:"confidence"`                     // 置信度 0-1
	Patterns            []PatternRecognitionResult `json:"patterns,omitempty"`             // 识别到的图形模式
	TechnicalIndicators *TechnicalIndicators       `json:"technical_indicators,omitempty"` // 技术指标数据
	Predictions         *PredictionResult          `json:"predictions,omitempty"`          // 预测数据
	Description         string                     `json:"description"`                    // 信号描述
	CreatedAt           time.Time                  `json:"created_at"`
	UpdatedAt           time.Time                  `json:"updated_at"`
}

// SignalCalculationRequest 信号计算请求
type SignalCalculationRequest struct {
	TSCode    string `json:"ts_code"`    // 股票代码
	Name      string `json:"name"`       // 股票名称
	StartDate string `json:"start_date"` // 开始日期
	EndDate   string `json:"end_date"`   // 结束日期
	Force     bool   `json:"force"`      // 是否强制重新计算
}

// SignalCalculationResponse 信号计算响应
type SignalCalculationResponse struct {
	Success bool         `json:"success"`
	Message string       `json:"message"`
	Signal  *StockSignal `json:"signal,omitempty"`
	Error   string       `json:"error,omitempty"`
}

// BatchSignalRequest 批量信号计算请求
type BatchSignalRequest struct {
	TSCodes []string `json:"ts_codes"` // 股票代码列表
	Force   bool     `json:"force"`    // 是否强制重新计算
}

// BatchSignalResponse 批量信号计算响应
type BatchSignalResponse struct {
	Total     int                         `json:"total"`   // 总数
	Success   int                         `json:"success"` // 成功数
	Failed    int                         `json:"failed"`  // 失败数
	Results   []SignalCalculationResponse `json:"results"` // 详细结果
	StartTime time.Time                   `json:"start_time"`
	EndTime   time.Time                   `json:"end_time"`
	Duration  string                      `json:"duration"`
}

// 新增技术指标结构定义

// WilliamsRIndicator 威廉指标 (%R)
type WilliamsRIndicator struct {
	WR14   JSONDecimal `json:"wr14"`   // 14日威廉指标
	Signal string      `json:"signal"` // 超买/超卖信号
}

// MomentumIndicator 动量指标
type MomentumIndicator struct {
	Momentum10 JSONDecimal `json:"momentum10"` // 10日动量
	Momentum20 JSONDecimal `json:"momentum20"` // 20日动量
	Signal     string      `json:"signal"`     // 买卖信号
}

// ROCIndicator 变化率指标
type ROCIndicator struct {
	ROC10  JSONDecimal `json:"roc10"`  // 10日变化率
	ROC20  JSONDecimal `json:"roc20"`  // 20日变化率
	Signal string      `json:"signal"` // 买卖信号
}

// ADXIndicator 平均方向指数
type ADXIndicator struct {
	ADX    JSONDecimal `json:"adx"`    // 平均方向指数
	PDI    JSONDecimal `json:"pdi"`    // 正方向指数
	MDI    JSONDecimal `json:"mdi"`    // 负方向指数
	Signal string      `json:"signal"` // 趋势强度信号
}

// SARIndicator 抛物线转向指标
type SARIndicator struct {
	SAR    JSONDecimal `json:"sar"`    // SAR值
	Signal string      `json:"signal"` // 买卖信号
}

// IchimokuIndicator 一目均衡表
type IchimokuIndicator struct {
	TenkanSen   JSONDecimal `json:"tenkan_sen"`    // 转换线
	KijunSen    JSONDecimal `json:"kijun_sen"`     // 基准线
	SenkouSpanA JSONDecimal `json:"senkou_span_a"` // 先行带A
	SenkouSpanB JSONDecimal `json:"senkou_span_b"` // 先行带B
	ChikouSpan  JSONDecimal `json:"chikou_span"`   // 滞后线
	Signal      string      `json:"signal"`        // 买卖信号
}

// ATRIndicator 平均真实范围
type ATRIndicator struct {
	ATR14  JSONDecimal `json:"atr14"`  // 14日ATR
	Signal string      `json:"signal"` // 波动率信号
}

// StdDevIndicator 标准差指标
type StdDevIndicator struct {
	StdDev20 JSONDecimal `json:"stddev20"` // 20日标准差
	Signal   string      `json:"signal"`   // 波动率信号
}

// HistoricalVolatilityIndicator 历史波动率
type HistoricalVolatilityIndicator struct {
	HV20   JSONDecimal `json:"hv20"`   // 20日历史波动率
	HV60   JSONDecimal `json:"hv60"`   // 60日历史波动率
	Signal string      `json:"signal"` // 波动率信号
}

// VWAPIndicator 成交量加权平均价
type VWAPIndicator struct {
	VWAP   JSONDecimal `json:"vwap"`   // 成交量加权平均价
	Signal string      `json:"signal"` // 买卖信号
}

// ADLineIndicator 累积/派发线
type ADLineIndicator struct {
	ADLine JSONDecimal `json:"ad_line"` // 累积/派发线
	Signal string      `json:"signal"`  // 买卖信号
}

// EMVIndicator 简易波动指标
type EMVIndicator struct {
	EMV14  JSONDecimal `json:"emv14"`  // 14日EMV
	Signal string      `json:"signal"` // 买卖信号
}

// VPTIndicator 量价确认指标
type VPTIndicator struct {
	VPT    JSONDecimal `json:"vpt"`    // 量价确认指标
	Signal string      `json:"signal"` // 买卖信号
}

File: web-react/src/types/stock.ts
/**
 * 股票相关数据类型定义 - Web版本
 * 对应Go后端的models包中的结构体
 */

// 股票基本信息
export interface StockBasic {
  ts_code: string;        // 股票代码
  symbol: string;         // 股票代码(不含后缀)
  name: string;           // 股票名称
  area?: string;          // 地域
  industry?: string;      // 行业
  market?: string;        // 市场类型
  list_date?: string;     // 上市日期
  is_hs?: string;         // 是否沪深港通
}

// 股票日线数据
export interface StockDaily {
  ts_code: string;        // 股票代码
  trade_date: string;     // 交易日期
  open: string;           // 开盘价
  high: string;           // 最高价
  low: string;            // 最低价
  close: string;          // 收盘价
  pre_close?: string;     // 前收盘价
  change?: string;        // 涨跌额
  pct_chg?: string;       // 涨跌幅
  vol: string;            // 成交量
  amount?: string;        // 成交额
}

// 技术指标数据 - 匹配后端Go结构
export interface TechnicalIndicators {
  ts_code: string;
  trade_date: string;
  
  // 传统指标
  macd?: MACDIndicator;
  rsi?: RSIIndicator;
  boll?: BollingerBandsIndicator;
  ma?: MovingAverageIndicator;
  kdj?: KDJIndicator;
  
  // 动量因子
  wr?: WilliamsRIndicator;
  momentum?: MomentumIndicator;
  roc?: ROCIndicator;
  
  // 趋势因子
  adx?: ADXIndicator;
  sar?: SARIndicator;
  ichimoku?: IchimokuIndicator;
  
  // 波动率因子
  atr?: ATRIndicator;
  stddev?: StdDevIndicator;
  hv?: HistoricalVolatilityIndicator;
  
  // 成交量因子
  vwap?: VWAPIndicator;
  ad_line?: ADLineIndicator;
  emv?: EMVIndicator;
  vpt?: VPTIndicator;
}

// MACD指标
export interface MACDIndicator {
  dif: number;
  dea: number;
  histogram: number;
  signal?: string;
}

// RSI指标
export interface RSIIndicator {
  rsi14: number;
  signal?: string;
}

// 布林带指标
export interface BollingerBandsIndicator {
  upper: number;
  middle: number;
  lower: number;
  signal?: string;
}

// 移动平均线指标
export interface MovingAverageIndicator {
  ma5: number;
  ma10: number;
  ma20: number;
  ma60: number;
  ma120: number;
}

// KDJ指标
export interface KDJIndicator {
  k: number;
  d: number;
  j: number;
  signal?: string;
}

// 威廉指标
export interface WilliamsRIndicator {
  wr14: number;
  signal?: string;
}

// 动量指标
export interface MomentumIndicator {
  momentum10: number;
  momentum20: number;
  signal?: string;
}

// 变化率指标
export interface ROCIndicator {
  roc10: number;
  roc20: number;
  signal?: string;
}

// ADX指标
export interface ADXIndicator {
  adx: number;
  pdi: number;
  mdi: number;
  signal?: string;
}

// SAR指标
export interface SARIndicator {
  sar: number;
  signal?: string;
}

// 一目均衡表
export interface IchimokuIndicator {
  tenkan_sen: number;
  kijun_sen: number;
  senkou_span_a: number;
  senkou_span_b: number;
  signal?: string;
}

// ATR指标
export interface ATRIndicator {
  atr14: number;
  signal?: string;
}

// 标准差指标
export interface StdDevIndicator {
  stddev20: number;
  signal?: string;
}

// 历史波动率指标
export interface HistoricalVolatilityIndicator {
  hv20: number;
  hv60: number;
  signal?: string;
}

// VWAP指标
export interface VWAPIndicator {
  vwap: number;
  signal?: string;
}

// A/D线指标
export interface ADLineIndicator {
  ad_line: number;
  signal?: string;
}

// EMV指标
export interface EMVIndicator {
  emv14: number;
  signal?: string;
}

// VPT指标
export interface VPTIndicator {
  vpt: number;
  signal?: string;
}

// 基本面数据
export interface FundamentalData {
  stock_basic?: StockBasic;
  income_statement?: IncomeStatement;
  balance_sheet?: BalanceSheet;
  cash_flow_statement?: CashFlowStatement;
  daily_basic?: DailyBasic;
}

// 利润表（简化版）
export interface IncomeStatement {
  ts_code: string;
  ann_date: string;       // 公告日期
  end_date: string;       // 报告期
  report_type: string;    // 报告类型
  basic_eps: number;      // 基本每股收益
  total_revenue: number;  // 营业总收入
  revenue: number;        // 营业收入
  total_profit: number;   // 利润总额
  n_income: number;       // 净利润
}

// 资产负债表（简化版）
export interface BalanceSheet {
  ts_code: string;
  ann_date: string;       // 公告日期
  end_date: string;       // 报告期
  total_share: number;    // 期末总股本
  total_assets: number;   // 资产总计
  total_liab: number;     // 负债合计
  total_hldr_eqy_inc_min_int: number; // 股东权益合计
}

// 现金流量表（简化版）
export interface CashFlowStatement {
  ts_code: string;
  ann_date: string;       // 公告日期
  end_date: string;       // 报告期
  n_cashflow_operate_a: number; // 经营活动现金流量净额
  n_cashflow_invest_a: number;  // 投资活动现金流量净额
  n_cash_flows_fnc_act: number; // 筹资活动现金流量净额
}

// 每日基本面数据
export interface DailyBasic {
  ts_code: string;
  trade_date: string;
  close?: number;         // 当日收盘价
  turnover_rate?: number; // 换手率
  pe?: number;           // 市盈率
  pb?: number;           // 市净率
  ps?: number;           // 市销率
  total_mv?: number;     // 总市值
  circ_mv?: number;      // 流通市值
}

// 收藏股票
export interface Favorite {
  id: string;           // 改为string以匹配后端UUID
  ts_code: string;
  name: string;
  start_date?: string;  // 添加开始日期
  end_date?: string;    // 添加结束日期
  group_id?: string;    // 改为string
  notes?: string;
  sort_order?: number;  // 改为sort_order
  created_at: string;
  updated_at: string;
}

// 添加收藏请求
export interface AddFavoriteRequest {
  ts_code: string;
  name: string;
  start_date?: string;
  end_date?: string;
  group_id?: string;
  notes?: string;
}

// 收藏分组
export interface FavoriteGroup {
  id: string;
  name: string;
  color?: string;       // 添加颜色
  sort_order?: number;
  created_at: string;
  updated_at: string;
}

// 创建分组请求
export interface CreateGroupRequest {
  name: string;
  color?: string;
}

// 更新分组请求
export interface UpdateGroupRequest {
  name?: string;
  color?: string;
  sort_order?: number;
}

// 更新收藏请求
export interface UpdateFavoriteRequest {
  start_date?: string;
  end_date?: string;
  group_id?: string;
  sort_order?: number;
}

// API响应格式
export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

// 分页响应
export interface PaginatedResponse<T = any> {
  total: number;
  page: number;
  page_size: number;
  data: T[];
}

// 搜索请求参数
export interface SearchParams {
  q: string;              // 搜索关键词
  limit?: number;         // 限制数量
  offset?: number;        // 偏移量
}

// 日期范围
export interface DateRange {
  start_date: string;     // 开始日期 YYYYMMDD
  end_date: string;       // 结束日期 YYYYMMDD
}

// 图表数据点
export interface ChartDataPoint {
  x: number | string;     // X轴数据
  y: number;              // Y轴数据
  date?: string;          // 日期字符串
}

// K线图数据
export interface CandlestickData extends ChartDataPoint {
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

// 买卖点预测
export interface TradingPointPrediction {
  type: 'BUY' | 'SELL';           // 买入或卖出
  price: number;                   // 预测价格
  date: string;                    // 预测日期 YYYYMMDD
  probability: number;             // 概率 (0-1)
  reason: string;                  // 预测理由
  indicators: string[];            // 相关指标
  signal_date: string;             // 信号产生的日期（基于哪一天的数据）
  backtested: boolean;             // 是否已回测
  is_correct?: boolean;            // 预测是否正确
  next_day_price?: number;         // 第二天收盘价
  price_diff?: number;             // 价格差值
  price_diff_ratio?: number;       // 价格差值百分比
}

// 预测结果
export interface PredictionResult {
  ts_code: string;
  trade_date: string;
  predictions: TradingPointPrediction[];
  confidence: number;              // 预测置信度 (0-1)
  updated_at: string;
}

// 形态模式摘要
export interface PatternSummary {
  ts_code: string;                 // 股票代码
  period: number;                  // 统计周期（天数）
  start_date: string;              // 开始日期 YYYYMMDD
  end_date: string;                // 结束日期 YYYYMMDD
  patterns: Record<string, number>; // 各种图形模式统计（形态名称 -> 出现次数）
  signals: Record<string, number>;  // 各种信号统计（信号类型 -> 出现次数）
  updated_at: string;              // 更新时间
}

// 收藏股票的信号数据
export interface FavoriteSignal {
  id: string;                      // 收藏ID
  ts_code: string;                 // 股票代码
  name: string;                    // 股票名称
  group_id?: string;               // 分组ID
  current_price: string;           // 当前价格
  trade_date: string;              // 交易日期
  indicators: TechnicalIndicators; // 技术指标
  predictions: TradingPointPrediction[]; // 买卖预测
  updated_at: string;              // 更新时间
}

// 收藏股票信号汇总响应
export interface FavoritesSignalsResponse {
  total: number;                   // 总数
  signals: FavoriteSignal[];       // 信号列表
  calculating?: boolean;           // 是否正在计算
  calculation_status?: {           // 计算状态
    status: string;
    message: {
      message: string;
    };
    detail: {
      is_calculating: boolean;
      completed: number;
      total: number;
      last_updated: string;
    };
  };
}